File: lc/coreprep/implementation/practice_iterations/patterns_iterations_1/all_paths.cpp
#include <algorithm>
#include <iostream>
#include <queue>
#include <sstream>
#include <string>
#include <vector>

using namespace std;

struct TreeNode {
  int val;
  TreeNode *left;
  TreeNode *right;

  TreeNode(int val) : val(val), left(nullptr), right(nullptr) {}
};

TreeNode *build(const vector<string> &a, int curr_idx) {
  if (curr_idx >= a.size() || a[curr_idx] == "null") {
    return nullptr;
  }

  TreeNode *node = new TreeNode(stoi(a[curr_idx]));
  node->left = build(a, 2 * curr_idx + 1);
  node->right = build(a, 2 * curr_idx + 2);

  return node;
}

TreeNode *build(const vector<string> &a) { return build(a, 0); }

void all_paths(TreeNode *root, vector<int> &curr_path,
               vector<vector<int>> &paths) {
  if (root == nullptr) {
    return;
  }

  if (root->left == nullptr && root->right == nullptr) {
    curr_path.push_back(root->val);
    paths.emplace_back(curr_path);
    curr_path.pop_back();
  }

  curr_path.push_back(root->val);
  all_paths(root->left, curr_path, paths);
  all_paths(root->right, curr_path, paths);
  curr_path.pop_back();
}

// O(nlogn) / O(n)
vector<vector<int>> all_paths(TreeNode *root) {
  vector<vector<int>> paths;
  vector<int> curr_path;

  all_paths(root, curr_path, paths);

  return paths;
}

int main() {
  ios::sync_with_stdio(0);
  cin.tie(0);

  int tt;
  string s;

  cin >> tt;
  cin.ignore(1, '\n');
  while (tt--) {

    getline(cin, s);

    stringstream ss(s);
    string tmp;
    vector<string> a;

    while (getline(ss, tmp, ' ')) {
      a.push_back(tmp);
    }

    TreeNode *root = build(a);
    auto res = all_paths(root);
    for (auto r : res) {
      for (int e : r) {
        cout << e << " ";
      }
      cout << "\n";
    }
  }
}

File: lc/coreprep/implementation/practice_iterations/patterns_iterations_2/AllPaths.java
import java.util.*;
import java.util.concurrent.*;
import java.util.function.*;

public class AllPaths {
  static class TreeNode {
    int val;
    TreeNode left;
    TreeNode right;

    TreeNode(int val) {
      this.val = val;
      this.left = null;
      this.right = null;
    }
  }

  private static TreeNode buildTree(String[] nodes) {
    return buildTree(0, nodes);
  }

  private static TreeNode buildTree(int currIdx, String[] nodes) {
    if (currIdx >= nodes.length || nodes[currIdx].equalsIgnoreCase("null")) {
      return null;
    }

    TreeNode node = new TreeNode(Integer.parseInt(nodes[currIdx]));
    node.left = buildTree(2 * currIdx + 1, nodes);
    node.right = buildTree(2 * currIdx + 2, nodes);

    return node;
  }

  public static void main(String[] args) {
    try (Scanner in = new Scanner(System.in)) {
      int tt = in.nextInt();
      in.nextLine();

      while (tt-- > 0) {
        String[] nodes = in.nextLine().trim().split(" ");

        TreeNode root = buildTree(nodes);
        var res = allPaths(root);
        for (var r : res) {
          for (int e : r) {
            System.out.printf("%d ", e);
          }
          System.out.println();
        }
      }
    }
  }

  // O(nlogn) / O(n)
  public static List<List<Integer>> allPaths(TreeNode root) {
    List<List<Integer>> allPaths = new ArrayList<>();
    allPaths(root, new ArrayList<>(), allPaths);

    return allPaths;
  }

  private static void allPaths(TreeNode root, List<Integer> currPath,
                               List<List<Integer>> allPaths) {
    if (root == null) {
      return;
    }

    currPath.add(root.val);

    if (root.left == null && root.right == null) {
      allPaths.add(new ArrayList<>(currPath));
    }

    allPaths(root.left, currPath, allPaths);
    allPaths(root.right, currPath, allPaths);
    currPath.remove(currPath.size() - 1);
  }
}